<!--
This launch file written by both Svena and Simon, contains launch instructions for both leader and follower robot
This is intended as a testing bed for the my_launch.launch files. The current version has the following components and functions:

Robot Follower has the following functions:
- Target Tracking (target_pose_tracking.py) & Tf Broadcasting (target_pose_broadcaster.py)
- Motion Controller (follower_controller.py)
- Visualization (trajectory - hector_trajectory_server, tf frames - robot_pose_ekf)

Robot Leader has the following functions:
- Motion Controller (leader_controller.py)
- Visualization (trajectory - hector_trajectory_server)

Global Nodes:
- Waypoints (waypoint_memory.py)
- Simulation Visualization (rviz)
-->

<launch>

  <arg name="waypoints_file" value="$(find smartcarts)/node/poses.csv"/>
  <param name="/camera_type" value="depth" /> <!--color or depth-->
  <param name="/ball_radius" value="0.04" />
  <param name="/debug_mode" value="1"/>
  <param name="min_follow_distance" value="0.7"/>
  <param name="min_camera_distance" value="0.6"/> <!--actual min = 0.5-->

  <include file="$(find gazebo_ros)/launch/empty_world.launch">
      <env name="GAZEBO_RESOURCE_PATH" value="$(find smartcarts)"/>
      <env name="GAZEBO_MODEL_PATH" value="$(find smartcarts)/models"/>
      <arg name="world_name" value="$(find smartcarts)/worlds/smartcarts.world"/>
      <arg name="gui" value="false"/>
  </include>

  <group ns="follower">
    <include file="$(find smartcarts)/launch/robot_follower.launch">
      <arg name="init_pose" value="-x -0.2 -y 0.5 -z 1.0 -R 0.0 -P 0.0 -Y 0.0" />
      <arg name="robot_name" value="follower" />
    </include>
    <!--Ball Tracking with Depth Camera Node-->
    <node pkg="smartcarts" type="target_pose_tracking.py" name="follower_ball_tracker"
        output="screen"/>
    <!--TF Broadcasting of Derived Leader's Waypoint Tracking-->
    <node pkg="smartcarts" type="target_pose_broadcaster.py" name="waypoint_broadcaster"
        output="screen"/>
   <node pkg="smartcarts" type="follower_controller.py" name="follower_controller"
        output="screen"/>
    <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="hector_trajectory_server"
        output="screen">
      <param name="target_frame_name" type="string" value="/world" />
      <param name="source_frame_name" type="string" value="/follower" />
      <param name="trajectory_update_rate" type="double" value="1" />
      <param name="trajectory_publish_rate" type="double" value="0.2" />
    </node>
  </group>

  <group ns="leader">
    <include file="$(find smartcarts)/launch/robot_leader.launch">
      <arg name="init_pose" value="-x 0.5 -y 0.5 -z 1.0 -R 0.0 -P 0.0 -Y 0.0" />
      <arg name="robot_name" value="leader" />
    </include>
<!--     <node pkg="smartcarts" type="line_follow.py" name="leader_line_follow" output="screen" />  -->
    <node pkg="smartcarts" type="leader_controller.py" name="my_leader" output="screen"/>
    <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="hector_trajectory_server" output="screen">
      <param name="target_frame_name" type="string" value="/world" />
      <param name="source_frame_name" type="string" value="/leader" />
      <param name="trajectory_update_rate" type="double" value="1" />
      <param name="trajectory_publish_rate" type="double" value="0.2" />
    </node>
  </group>

  <node pkg="smartcarts" type="waypoint_memory.py" name="waypoint_memory"
    output="screen"/>

  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find smartcarts)/rviz/follower_pathview.rviz" />

</launch>
